saige_assoc_set_test.py

```diff
@@ -364,23 +388,37 @@ def manage_concurrency_for_job(job: hb.batch.job.Job):
                     gene_dependency = null_job
 
                 # step 2 (cis eQTL set-based test)
-                step2_job, step2_output = build_run_set_based_test_command(
-                    set_output_path=output_path(
-                        f'{celltype}/{chromosome}/{celltype}_{gene}_cis_set', 'analysis'
-                    ),
-                    vcf_file=vcf_file_path,
+                # unique key for this set-based test
+                set_key = f'{celltype}/{chromosome}/{celltype}_{gene}_cis_set'
+                # unique output path for this set-based test
+                set_output_path = output_path(set_key, 'analysis')
+
+                # if the output exists, do nothing
+                if to_path(set_output_path).exists():
+                    continue
+
+                # instruct an additional command to run inside this VM
+                build_run_set_based_test_command(
+                    job=step2_job,
+                    set_key=set_key,
+                    set_output_path=set_output_path,
+                    vcf_group=vcf_group,
                     chrom=(chromosome[3:]),
                     group_file=group_path,
                     gmmat_model_path=null_output['rda'],
                     variance_ratio_path=null_output['varianceRatio.txt'],
                 )
-                if step2_job is not None:
-                    step2_job.depends_on(gene_dependency)
-                    gene_dependency = step2_job
+                jobs_in_vm += 1
+
+                # if we ran this job, an additional dependency
+                step2_job.depends_on(gene_dependency)
 
                 # add this job to the list of jobs for this cell type
-                if gene_dependency is not None:
-                    celltype_jobs.setdefault(celltype, []).append(gene_dependency)
+                celltype_jobs.setdefault(celltype, []).append(step2_job)
+
+                # check if we need a new VM, i.e. we've hit the jobs-per-VM limit
+                if jobs_in_vm >= jobs_per_vm:
+                    step2_job = create_a_2b_job()
```

Comment:
Just feel like we should create a new 2b job here and then set `jobs_in_vm` back to `0`? Hope I'm following the code correctly.

Reply:
yup, I forgot to reset the job counter